### PR TITLE
Eager `t.*` access in `shouldStoreRHSInTemporaryVariable`

### PR DESCRIPTION
**Pull request by zhangenming**
use more deconstruction.

| Q                        | A 
| ------------------------ | ---
| Fixed Issues?            |
| Patch: Bug Fix?          |
| Major: Breaking Change?  |
| Minor: New Feature?      |
| Tests Added + Pass?      | 
| Documentation PR Link    | 
| Any Dependency Changes?  |
| License                  | MIT

<a href="https://gitpod.io/#https://github.com/babel/babel/pull/14191"></a>

### CODE REVIEW

File: packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts [hunk 1]
```diff
@@ -1,35 +1,35 @@
...
 /**
  * This is a helper function to determine if we should create an intermediate variable
  * such that the RHS of an assignment is not duplicated.
  *
  * See https://github.com/babel/babel/pull/13711#issuecomment-914388382 for discussion
  * on further optimizations.
  */
-export default function shouldStoreRHSInTemporaryVariable(node: t.LVal) {
-  if (t.isArrayPattern(node)) {
+export default function shouldStoreRHSInTemporaryVariable(node: LVal) {
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts:11]: `LVal` is a type export. We'd better leave it as-is.

File: packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts [hunk 2]
```diff
@@ -1,35 +1,35 @@
...
+  } else if (isObjectPattern(node)) {
     const { properties } = node;
     if (properties.length > 1) return true;
     else if (properties.length === 0) return false;
     else {
       const firstProperty = properties[0];
       if (isObjectProperty(firstProperty)) {
         // the value of the property must be an LVal
-        return shouldStoreRHSInTemporaryVariable(firstProperty.value as t.LVal);
+        return shouldStoreRHSInTemporaryVariable(firstProperty.value as LVal);
```
Comment (JLHwung) [inline packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts:24]: ditto.

**Comment by JLHwung** [comment]
@zhangenming Can you fix the lint warnings? You can fix them by running `make fix-js`.

File: packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts [hunk 3]
```diff
@@ -1,6 +1,6 @@
 import { types as t } from "@babel/core";
 
-const { isObjectProperty } = t;
+const { isObjectProperty, isArrayPattern, isObjectPattern, isAssignmentPattern, isRestElement, isIdentifier } = t;
```
Comment (The-x-Theorist) [inline packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts:3]:
```suggestion
const {
    isObjectProperty,
    isArrayPattern,
    isObjectPattern,
    isAssignmentPattern,
    isRestElement,
    isIdentifier
} = t;
```

Comment (The-x-Theorist) [reply packages/babel-plugin-proposal-object-rest-spread/src/shouldStoreRHSInTemporaryVariable.ts:3]: Changing the deconstruction to this structure might solve the linting error in CI.

**Review by The-x-Theorist — CHANGES_REQUESTED** [review]

**Review by nicolo-ribaudo — APPROVED** [review]
Thanks!

**Pull request merged**